File: app/backend/src/application/dto/product_dto.rs
/// ProductのViewModel（Application層）
/// CQRS Query側専用：Infrastructure層から直接構築
/// パフォーマンス重視でドメインモデルを経由しない
#[derive(Debug, Clone)]
pub struct ProductDTO {
    pub id: String,
    pub name: String,
    pub images: Vec<String>,
    pub category: String,
    pub description: String,
    pub is_best_seller: bool,
    pub is_quick_ship: bool,
    pub variants: Vec<VariantDTO>,
}

/// Product variantのビューモデル（Application層）
#[derive(Debug, Clone)]
pub struct VariantDTO {
    pub id: String,
    pub sku_code: String,
    pub name: String,
    pub color: String,
    pub material: String,
    pub dimensions: String,
    pub price: u32,
    pub sale_price: Option<u32>,
    pub stock_quantity: u32,
    pub reserved_quantity: u32,
    pub display_order: u32,
    pub image: Option<String>,
    pub is_on_sale: bool,
    pub is_sold_out: bool,
}


impl ProductDTO {}

impl VariantDTO {
    /// Infrastructure層から直接構築用
    pub fn new(
        id: String,
        sku_code: String,
        name: String,
        color: String,
        material: String,
        dimensions: String,
        price: u32,
        sale_price: Option<u32>,
        stock_quantity: u32,
        reserved_quantity: u32,
        display_order: u32,
        image: Option<String>,
    ) -> Self {
        let is_on_sale = sale_price.is_some();
        let is_sold_out = stock_quantity == 0;

        Self {
            id,
            sku_code,
            name,
            color,
            material,
            dimensions,
            price,
            sale_price,
            stock_quantity,
            reserved_quantity,
            display_order,
            image,
            is_on_sale,
            is_sold_out,
        }
    }
}

File: app/backend-ts/src/application/dto/product.dto.ts
export class VariantDto {
  constructor(
    public readonly id: string,
    public readonly skuCode: string,
    public readonly name: string,
    public readonly color: string,
    public readonly material: string,
    public readonly dimensions: string,
    public readonly price: number,
    public readonly salePrice: number | null,
    public readonly stockQuantity: number,
    public readonly reservedQuantity: number,
    public readonly displayOrder: number,
    public readonly image: string | null,
    public readonly isOnSale: boolean,
    public readonly isSoldOut: boolean,
  ) {}
}

export class ProductDto {
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly images: string[],
    public readonly category: string,
    public readonly description: string,
    public readonly isBestSeller: boolean,
    public readonly isQuickShip: boolean,
    public readonly variants: VariantDto[],
  ) {}
}

export class ProductSummaryDto {
  constructor(
    public readonly id: string,
    public readonly name: string,
    public readonly category: string,
    public readonly basePrice: number,
    public readonly salePrice: number | null,
    public readonly image: string | null,
    public readonly colors: string[],
    public readonly isBestSeller: boolean,
    public readonly isQuickShip: boolean,
    public readonly stockQuantity: number,
  ) {}
}

export class ProductListDto {
  constructor(
    public readonly products: ProductSummaryDto[],
    public readonly totalCount: number,
    public readonly page: number,
    public readonly perPage: number,
    public readonly hasNextPage: boolean,
    public readonly hasPreviousPage: boolean,
  ) {}
}
